zerver/tornado/handlers.py

```diff
@@ -96,6 +97,28 @@ def __repr__(self) -> str:
         descriptor = get_descriptor_by_handler_id(self.handler_id)
         return "AsyncDjangoHandler<%s, %s>" % (self.handler_id, descriptor)
 
+    def _legacy_get_response(self, request: HttpRequest) -> HttpResponse:
+        """
+        Apply process_request() middleware and call the main _get_response(),
+        if needed. Used only for legacy MIDDLEWARE_CLASSES.
+        """
+        response = None
+        # Apply request middleware
+        for middleware_method in self._request_middleware:
+            response = middleware_method(request)
+            if response:
+                break
+
+        if response is None:
+            response = self._get_response(request)
+        return response
+
+    def handle_uncaught_exception(self, request: HttpRequest,
+                                  resolver: resolvers.RegexURLResolver,
+                                  exc_info: Any) -> HttpResponse:
+        """Allow subclasses to override uncaught exception handling."""
+        return handle_uncaught_exception(request, resolver, exc_info)
+
```

Comment:
So, most of this file is copied from Django; usually what we do when updating for a major Django release is re-sync the file (since the main change that we have made is adding the `RespondAsynchronously` feature, though probably at this point there's a bunch of type annotation stuff that makes it harder to read that, and there might be a bit more).

Reply:
We need to copy. Django 2.0 doesn't have these functions.

Reply:
For ref: https://github.com/django/django/commit/d334f46b7a080fd3eb720141c19b37b10704a352

https://code.djangoproject.com/ticket/26601

Reply:
That's not what I'm saying.  What I'm saying is that for migrating this code to Django 2.0, we should re-copy the `django.core.handlers.base.get_response` function from Django 2.0 (and then re-apply our small edits), so that we're doing the correct thing in Django 2.0.  There may be other Django 2.0 migrations that we need to apply.

Reply:
I see. Let me check that.

Reply:
This would require some re-thinking. Our code relies on middleware short-circuiting, whereas the new-style middlewares don't do that.